mcstatus/status_response.py

```diff
@@ -0,0 +1,317 @@
+
+__all__ = [
+    "STYLE_MAP",
+    "AbstractDataclass",
+    "MCStatusResponse",
+    "JavaStatusResponse",
+    "BedrockStatusResponse",
+    "StatusPlayers",
+    "JavaStatusPlayers",
+    "BedrockStatusPlayers",
+    "JavaStatusPlayer",
+    "StatusVersion",
+    "JavaStatusVersion",
+    "BedrockStatusVersion",
+]
+
+STYLE_MAP = {
+    "bold": "l",
+    "italic": "o",
+    "underlined": "n",
+    "obfuscated": "k",
+    "color": {
+        "dark_red": "4",
+        "red": "c",
+        "gold": "6",
+        "yellow": "e",
+        "dark_green": "2",
+        "green": "a",
+        "aqua": "b",
+        "dark_aqua": "3",
+        "dark_blue": "1",
+        "blue": "9",
+        "light_purple": "d",
+        "dark_purple": "5",
+        "white": "f",
+        "gray": "7",
+        "dark_gray": "8",
+        "black": "0",
+    },
+}
+
+
+def _validate_data(raw: Dict[str, Any], who: str, required: Iterable[Tuple[str, type]]) -> None:
+    """Ensure that all required keys are present, and have the specified type.
+
+    :param raw: The raw dict answer to check.
+    :param who: The name of the object that is checking the data. Example `status`, `player` etc.
+    :param required: An iterable of string and type. The string is the required key which must be in `raw`, and
+        the `type` is the type that the key must be. If you want to ignore check of the type, set the type to `object`.
+    :raises ValueError: If the required keys are not present.
+    :raises TypeError: If the required keys are not of the specified type.
+    """
+    for required_key, required_type in required:
+        if required_key not in raw:
+            raise ValueError(f"Invalid {who} object (no '{required_key}' value)")
+        if not isinstance(raw[required_key], required_type):
+            raise TypeError(
+                f"Invalid {who} object (expected '{required_key}' to be {required_type}, was {type(raw[required_key])})"
+            )
+
+
+@dataclass
+class AbstractDataclass(ABC):
+    """Abstract base class for all dataclasses.
+
+    We can't use just `ABC` because we need raise TypeError if we try to
+    instantiate an abstract dataclass. Class which inherit from `ABC` raises
+    it only if it has abstract methods.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if cls is AbstractDataclass or cls.__bases__[0] is AbstractDataclass:
+            raise TypeError("Cannot instantiate abstract class.")
+        return super().__new__(cls)
+
+
+@dataclass
+class MCStatusResponse(AbstractDataclass):
+    """Class for storing shared data from a status response.
+
+    :param motd: Message Of The Day. Also known as `Description`.
+    :param latency: Latency between a server and the client (you). In milliseconds.
+    """
+
+    players: StatusPlayers
+    version: StatusVersion
+    motd: str
+    latency: float
+
+    @classmethod
+    @abstractmethod
+    def build(cls, *args, **kwargs) -> MCStatusResponse:
```

Comment:
Let's use `typing_extensions.Self` instead of defining the exact class as the return type of `build` method.

This is benefitial, because when a class is subclassed, if it inherits this `build` method without explicitly overriding it, even though calling it would produce an instance of that subclass, it would be downcasted to the parent class, as defined here and type-checkers would treat it as such. For more info, check [PEP 673](https://peps.python.org/pep-0673/)

Note that `typing_extensions` is not a runtime dependency, which means it should only be imported on typing time. You can achieve this by using `if typing.TYPE_CHECKING`. Along with this, we prefer to use `from future import __annotations__` to avoid having to use strings with these typying time only imports (without this import: `-> "Self"` , with \_\_annotations\_\_: `-> Self`).

you will need to import 

This applies for all `build` methods (or other constructor methods) in all of the changed code.

Reply:
`typing_extensions` is runtime dependece, check `poetry.lock`.

Reply:
Used in 43438e3a9c23361838ca8c34ce8adceae0afb409.